### PR TITLE
PERF-5184 Add workload reproducing an OOM on mongos caused by a cursor storm

### PR DESCRIPTION
**Pull request by robsndr**
Jira Ticket: [PERF-5184](https://jira.mongodb.org/browse/PERF-5184)

Whats Changed:
Add a new workload reproducing an OOM on mongos caused by a cursor storm.

Workload Submission form:
If applicable, only required if is a new workload being added. Form can be found [here](https://docs.google.com/forms/d/1r0kOHvFUa7rJxVmX_wp9prxYU5K155yKyZ1y3d5yhZg/edit)

### CODE REVIEW

File: src/workloads/scale/CursorStormMongos.yml [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+      NopInPhasesUpTo: *MaxPhases
+      PhaseConfig:
+        Repeat: *SortRepeat
+        Database: *DBName
+        Operations:
+        - OperationMetricsName: SortMany
+          OperationName: RunCommand
+          OperationCommand:
+            # Loader default collection name.
+            aggregate: Collection0
```
Comment (josefahmad) [inline src/workloads/scale/CursorStormMongos.yml:158]: Can we simply stream results from the shard, e.g. a find resulting in a plain index scan or collection scan. We don't need to tax the shards, all we need is to generate a batch to feed the mongos.

File: src/workloads/scale/CursorStormMongos.yml [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+                  {$let:
+                    {vars:
+                      {start: {^RandomInt: {min: 0, max: *NumDocs}}},
+                    in: {$and: [$gte: ["$a", "$$start"], $lt: ["$a", {$add: ["$$start", *SortStep]}]]}
+                    }
+                  }
+                }
+              },
+              # Merge-sorting results imposes additional memory pressure on mongos
+              {$sort: {b: 1}}]
```
Comment (josefahmad) [inline src/workloads/scale/CursorStormMongos.yml:171]: Wasn't aware of this. How so?

**Review by josefahmad — COMMENTED** [review]
Thanks for putting up this PR. I think we can simplify the workload on the shard, in that all I need is shards returning results to the mongos.

File: src/workloads/scale/CursorStormMongos.yml (hunk 1 above)
Comment (robsndr) [reply src/workloads/scale/CursorStormMongos.yml:158]: I've kept the sort based on the observation that without it, the workload takes ~2x longer to reach OOM, with ~2x more cursors being opened on the mongos. Could it be that using the sort-stage affects how data is fetched/buffered on the mongos? 

Shall we simplify the workload regardless?

File: src/workloads/scale/CursorStormMongos.yml (hunk 2 above)
Comment (robsndr) [reply src/workloads/scale/CursorStormMongos.yml:171]: Please see above.

File: src/workloads/scale/CursorStormMongos.yml [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+  # The size of documents to process in a given step should approach 16MB per shard. Therefore, the
+  # number of documents to sort per shard is tuned such that SortStepPerShard * DocSize -> 16 MB.
+  SortStepPerShard: &SortStepPerShard 90
+  # With data distributed evenly across shards, the total size of documents to process in a
+  # given step should be equall to NumOfShards * SortStepPerShard.
+  SortStep: &SortStep
+    ^NumExpr:
+      withExpression: "num_shards * sort_step_per_shard"
+      andValues: {num_shards: *NumShards, sort_step_per_shard: *SortStepPerShard}
```
Comment (josefahmad) [inline src/workloads/scale/CursorStormMongos.yml:59]: IIUC all of this is to set a specific batch size. Why don't we just use the default batch size?

File: src/workloads/scale/CursorStormMongos.yml [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+Description: |
+  Used to test performance of the router under memory pressure caused by accumulating
+  many heavy cursors. The workload is expected to fail due to host(s) being unreachable as a
+  result of mongos running out of memory.
+
+  To achieve this, many threads are spawned to run aggregate sort on a number of documents.
+  The document size and the number of documents to sort by each thread are tuned so, that the
+  operations do not spill to disk [<100MB], and mongos is able to exhaust the cursors on shards
+  when pre-filling its buffers [<16MB per shard]. As a result, memory pressure on the shards
+  remains low, while it's kept large on the mongos.
```
[the remark this answers is not in the document]
Comment (robsndr) [reply src/workloads/scale/CursorStormMongos.yml]: Got rid of the sort, as it didn't have its former effect on the workload after removing the aggregation.

File: src/workloads/scale/CursorStormMongos.yml (hunk 1 above)
Comment (robsndr) [reply src/workloads/scale/CursorStormMongos.yml]: I've assumed that we might see some effects related to caching if the exact same query is run by many threads, many times. However, this seems to not be the case! Removed some noise from the workload.

> Can we simplify this workload by doing a simple unfiltered find + sort stage on an index and avoiding a blocking sort on the shards?

Updated.

> Out of curiosity, how long does it take for this stage to OOM with and without the sort stage?

It was ~40s vs ~80s. However, now its ~20s, and I could barely see any difference with or without the sort. Therefore, I've just removed the sort altogether.

Comment (josefahmad) [reply src/workloads/scale/CursorStormMongos.yml]: Perhaps the speed-up is due to https://jira.mongodb.org/browse/DEVPROD-4593? In any case, I'm happy without the sort stage!

**Review by josefahmad — APPROVED** [review]
LGTM! Thanks Robert, and sorry for the delay in reviewing

File: src/workloads/scale/CursorStormMongos.yml [hunk 5]
```diff
@@ -0,0 +1,147 @@
+SchemaVersion: 2018-07-01
+Owner: Service Architecture
```
Comment (josefahmad) [inline src/workloads/scale/CursorStormMongos.yml:2]: @robsndr I just come across [this slack thread|https://mongodb.slack.com/archives/C0685CU3EPJ/p1708101747222789] suggesting the owner should be Query.

Comment (robsndr) [reply src/workloads/scale/CursorStormMongos.yml]: Changed the owning team to Query.

**Review by josefahmad — APPROVED** [review]
LGTM!

**Comment by robsndr** [comment]
evergreen merge

**Pull request merged**